manager/allocator/allocator_test.go

```diff
@@ -716,6 +718,139 @@ func isValidSubnet(t assert.TestingT, subnet string) bool {
 	return assert.NoError(t, err)
 }
 
+func TestCNIAllocator(t *testing.T) {
+	s := store.NewMemoryStore(nil)
+	assert.NotNil(t, s)
+	defer s.Close()
+
+	a, err := New(s, cni.New())
+	assert.NoError(t, err)
+	assert.NotNil(t, a)
+
+	// Try adding some objects to store before allocator is started
+	assert.NoError(t, s.Update(func(tx store.Tx) error {
+		n1 := &api.Network{
+			ID: "cniTestID1",
+			Spec: api.NetworkSpec{
+				DriverConfig: &api.Driver{
+					Name: "cni",
+					Options: map[string]string{
+						"Config": `{
+  "cniVersion": "0.3.0",
+  "name": "default",
+  "type": "bridge",
+  "bridge": "cni0",
+  "isGateway": true,
+  "ipam": {
+    "type": "host-local",
+    "subnet": "10.1.0.0/16",
+    "gateway": "10.1.0.1"
+  },
+  "dns": {
+    "nameservers": [ "10.1.0.1" ]
+  }
+}`},
+				},
+				Annotations: api.Annotations{
+					Name: "cniTest1",
+				},
+			},
+		}
+
+		assert.NoError(t, store.CreateNetwork(tx, n1))
+		s1 := &api.Service{
+			ID: "cniTestServiceID1",
+			Spec: api.ServiceSpec{
+				Annotations: api.Annotations{
+					Name: "cniService1",
+				},
+				Task: api.TaskSpec{
+					Networks: []*api.NetworkAttachmentConfig{
+						{
+							Target: "cniTestID1",
+						},
+					},
+				},
+				//Endpoint: &api.EndpointSpec{
+				//	Mode: api.ResolutionModeVirtualIP,
+				//	Ports: []*api.PortConfig{
+				//		{
+				//			Name:          "portName",
+				//			Protocol:      api.ProtocolTCP,
+				//			TargetPort:    8000,
+				//			PublishedPort: 8001,
+				//		},
+				//	},
+				//},
+			},
+		}
+		assert.NoError(t, store.CreateService(tx, s1))
+
+		t1 := &api.Task{
+			ID: "cniTestTaskID1",
+			Status: api.TaskStatus{
+				State: api.TaskStateNew,
+			},
+			Networks: []*api.NetworkAttachment{
+				{
+					Network: n1,
+				},
+			},
+		}
+		assert.NoError(t, store.CreateTask(tx, t1))
+
+		t2 := &api.Task{
+			ID: "cniTestTaskIDPreInit",
+			Status: api.TaskStatus{
+				State: api.TaskStateNew,
+			},
+			ServiceID:    "cniTestServiceID1",
+			DesiredState: api.TaskStateRunning,
+		}
+		assert.NoError(t, store.CreateTask(tx, t2))
+		return nil
+	}))
+
+	netWatch, cancel := state.Watch(s.WatchQueue(), api.EventUpdateNetwork{}, api.EventDeleteNetwork{})
+	defer cancel()
+	taskWatch, cancel := state.Watch(s.WatchQueue(), api.EventUpdateTask{}, api.EventDeleteTask{})
+	defer cancel()
```

Comment:
I believe this works as intended, but it might clarify the code to avoid reusing the name `cancel`.